resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```diff
@@ -110,117 +104,381 @@ private[spark] class Client(
+      driverService: Service) = {
+    sparkConf.getOption("spark.app.id").foreach { id =>
+      logWarning(s"Warning: Provided app id in spark.app.id as $id will be" +
+        s" overridden as $kubernetesAppId")
+    }
+    sparkConf.set(KUBERNETES_DRIVER_POD_NAME, kubernetesAppId)
+    sparkConf.set(KUBERNETES_DRIVER_SERVICE_NAME, driverService.getMetadata.getName)
+    sparkConf.set("spark.app.id", kubernetesAppId)
+    sparkConf.setIfMissing("spark.app.name", appName)
+    sparkConf.setIfMissing("spark.driver.port", DEFAULT_DRIVER_PORT.toString)
+    sparkConf.setIfMissing("spark.blockmanager.port",
+      DEFAULT_BLOCKMANAGER_PORT.toString)
+    val driverSubmitter = buildDriverSubmissionClient(kubernetesClient, driverService,
+      driverSubmitSslOptions)
+    // Sanity check to see if the driver submitter is even reachable.
+    driverSubmitter.ping()
+    logInfo(s"Submitting local resources to driver pod for application " +
+      s"$kubernetesAppId ...")
+    val submitRequest = buildSubmissionRequest()
+    driverSubmitter.submitApplication(submitRequest)
+    logInfo("Successfully submitted local resources and driver configuration to" +
+      " driver pod.")
+    // After submitting, adjust the service to only expose the Spark UI
+    val uiServicePort = new ServicePortBuilder()
+      .withName(UI_PORT_NAME)
+      .withPort(uiPort)
+      .withNewTargetPort(uiPort)
+      .build()
+    kubernetesClient.services().withName(kubernetesAppId).edit().editSpec()
+      .withType("ClusterIP")
+      .withPorts(uiServicePort)
+      .endSpec()
+      .done()
+    logInfo("Finished submitting application to Kubernetes.")
+  }
+
+  private def launchDriverKubernetesComponents(
+      kubernetesClient: KubernetesClient,
+      parsedCustomLabels: Map[String, String],
+      submitServerSecret: Secret,
+      driverSubmitSslOptions: SSLOptions,
+      sslSecrets: Array[Secret],
+      sslVolumes: Array[Volume],
+      sslVolumeMounts: Array[VolumeMount],
+      sslEnvs: Array[EnvVar],
+      isKeyStoreLocalFile: Boolean): (Pod, Service) = {
+    val endpointsReadyFuture = SettableFuture.create[Endpoints]
+    val endpointsReadyWatcher = new DriverEndpointsReadyWatcher(endpointsReadyFuture)
+    val serviceReadyFuture = SettableFuture.create[Service]
+    val serviceReadyWatcher = new DriverServiceReadyWatcher(serviceReadyFuture)
+    val podReadyFuture = SettableFuture.create[Pod]
+    val podWatcher = new DriverPodReadyWatcher(podReadyFuture)
+    val (readyPod, readyService) = Utils.tryWithResource(kubernetesClient
+        .pods()
+        .withName(kubernetesAppId)
+        .watch(podWatcher)) { _ =>
+      Utils.tryWithResource(kubernetesClient
+          .services()
+          .withName(kubernetesAppId)
+          .watch(serviceReadyWatcher)) { _ =>
+        Utils.tryWithResource(kubernetesClient
+            .endpoints()
+            .withName(kubernetesAppId)
+            .watch(endpointsReadyWatcher)) { _ =>
```

Comment:
@lins05 the latest patch rearranges the logic a little bit - how does it look now? I don't think extracting the three `tryWithResource` is the real problem, so much as having this method do both the Kubernetes component creation and the adjustment for owner references seems like a bit of overload.

I think the idea of having a `tryWithResource` that takes multiple Closeable factories is intriguing - the tricky part would be having the closure accept a variable-length argument list, if that makes sense.